### PR TITLE
Add runAfterMutation method to mutateAndGetPayload

### PR DESCRIPTION
**Pull request by jamesone**
Note: The linter errors will need to be fixed (this can be done once I get confirmation that this feature is something that could be implemented). I will also need to update the example `.js` file.

I was surprised to find out that this method was not already there? There are multiple use cases for this method:

1. Caching - If you want to cache specific payloads
2. Realtime - If you aren't using graphql subscriptions and have built your own custom solution, being able to emit the final payload would be SUPER handy. For instance I had to do a lot of manual work to replicate a graphql payload (manually converting to global ids, fetching database objects which aren't in the current mutation scope but would be in the resolved payload, etc...) 
3. Debugging 
4. If you need to run methods that require access to the mutation payload, you can add them in here.

A simple sample would go as follows:

```
mutateAndGetPayload...,
runAfterMutation: payload => new Promise((resolve) => {
    redis.set('updated_user', payload.user);
    resolve();
})
```

Is this a feature you would consider having?

### CODE REVIEW

**Comment by jamesone** [comment]
Fixed flow errors ✅ 

@wincent Could you maybe take a look at this PR and add some feedback? I'm pretty certain that this would be a useful feature to have for a lot of people 😄 The good thing is that it's optional 👍

File: src/mutation/mutation.js [hunk 1]
```diff
@@ -105,6 +115,19 @@ export function mutationWithClientMutationId(
...
+          if (runAfterMutation) {
+            return Promise.resolve(runAfterMutation(payload))
+              .then(() => payload)
+              .catch(err => {
+                /**
+                 * This shouldn't make the mutation fail. However
+                 * we should let the user know about there error
+                 * & then return the payload
+                 */
+                console.log(err, 'Error in runAfterMutation');
```
Comment (jamesone) [inline src/mutation/mutation.js]: console.log gives a linter warning, but I think in this case we should at least let the user know that something in there runAfterMutation method failed?

**Comment by jamesone** [comment]
@leebyron Hey 😄 When you have the time could you go over this feature?

**Comment by sibelius** [comment]
can't if do this logic inside `mutateAndGetPayload`?

**Comment by jamesone** [comment]
@sibelius Do you mean that it can already be done (i don't think it can) or the logic can be moved into the mutateAndGetPayload?

**Comment by idris** [comment]
@jamesone why not just put your "after" code at the end of `mutateAndGetPayload`?  I do that a lot, and just fire-and-forget asynchronous stuff (don't `await`, so the mutation can return to the client)

**Comment by jamesone** [comment]
@idris But how would you access the **resolved output types**? Can you do that inside the **mutateAndGetPayload**? Maybe I missed something? However when I was doing this, the source code suggested that I am unable to access the **resolved** output types inside the **mutateAndGetPayload**?

**Comment by idris** [comment]
I don't know what you mean by the *resolved output types*.  Presumably whatever code your resolver is using, you can use in your `mutateAndGetPayload` as well.

Also, the `payload` is just the return value from `mutateAndGetPayload`, so you obviously have access to that.

**Comment by jamesone** [comment]
Oh, so let's say I add a UserType to my mutation output type, the client has access to all the resolved fields in that user type, so what I'm wanting to do is access that user type and all its resolved fields. Which is why I added the runAfter method. Does that make sense?


Sent from Outlook<https://aka.ms/qtex0l>
________________________________

**Comment by idris** [comment]
@jamesone I think you're confusing client side and server side code.  What gets resolved is sent back to the client.  The server code never really deals with resolved objects.  Instead, your server code is dealing with your own models.

So, for example, if you have a `User` with a `Post`, you'd want to do something like `user.getPost()` on your models to get a Post.  If you are afraid that your resolvers will cause `Post` to be queried and then your "after" code will also cause `Post` to be queried, you can look into `DataLoader` to make that query only happen once.

**Comment by taion** [comment]
On accessing fields, that's really a broader issue with the GraphQL.js API, and I don't think it's quite right to work around it just for this one method.

Really the GraphQL.js API is a schema/serialization layer – think of it as a fancy version of joi (or yup), in that the serialization helpers don't see each other's output; they work instead on the underlying model.

So this sort of thing is really better handled in a separate model layer, IMO.

**Comment by jamesone** [comment]
@taion Yeah I understand that 👍 

However I still think this command could come in handy & that's after coming back and looking at this PR for the first time in a while 😄 

Maybe we could make it so you cannot mutate the `payload` that gets passed to `runAfterMutation`...instead you can just use it for stuff like caching 🤔

**Comment by taion** [comment]
Those sort of use cases don't need to be handled within GraphQL execution, though. You can move them one level up into your GraphQL server.

**Comment by jamesone** [comment]
@taion 👍

**Comment by IvanGoncharov** [comment]
Closing, see @taion excellent explaination why it should be part of GraphQL layer.

**Pull request closed without merging**